source/agora/consensus/protocol/Nominator.d

```diff
@@ -466,6 +441,48 @@ extern(D):
             log.info("Rejected invalid envelope: {}", scpPrettify(&envelope));
     }
 
+    /***************************************************************************
+
+        Store the latest SCP sate for restore
+
+    ***************************************************************************/
+
+    public void storeLatestState () @safe nothrow
+    {
+        vector!SCPEnvelope envelopes;
+
+        () @trusted
+        {
+            if (this.scp.empty())
+                return;
+            envelopes = this.scp.getLatestMessagesSend(this.scp.getHighSlotIndex());
+        }();
+
+        // Clean the previous envelopes from the DB
+        this.scp_envelope_store.removeAll();
```

Comment:
I think this and the adding needs to be atomic with `db.begin` / `db.commit`. If removing succeeds but adding new envelopes fails, then we will have lost the old envelope messages.

Reply:
That's a good point.